File: 01/p1.rs
// The natural numbers below 10 that are multiples of 3 or 5 are 3, 5, 6 and 9.
// The sum of these multiples is 23.
// Find the sum of all the multiples of 3 or 5 below 1000.

fn sum_under(x: u32) -> u32 {
    ((x * x) + x) / 2
}

fn sum_steps(step: u32, limit: u32) -> u32 {
    step * sum_under((limit - 1) / step)
}

fn main() {
    let a = sum_steps(3, 1000);
    let b = sum_steps(5, 1000);
    let c = sum_steps(15, 1000);
    println!("{}", a + b - c)
}
// 233168

File: 01/p1.exs
# The natural numbers below 10 that are multiples of 3 or 5 are 3, 5, 6 and 9.
# The sum of these multiples is 23.
# Find the sum of all the multiples of 3 or 5 below 1000.

defmodule E do

  def sum_under(x) do
    ((x * x) + x) / 2
  end

  def sum_steps(step, limit) do
    step * sum_under(div(limit - 1, step))
  end

end

s = &E.sum_steps(&1, 1000)

IO.puts(s.(3) + s.(5) - s.(15))
# 233168.0
